xonsh/aliases.py

```diff
@@ -86,37 +93,27 @@ def source_foreign(args, stdin=None):
                             aliascmd=ns.aliascmd, extra_args=ns.extra_args,
                             safe=ns.safe, prevcmd=ns.prevcmd,
                             postcmd=ns.postcmd, funcscmd=ns.funcscmd,
-                            sourcer=ns.sourcer)
+                            sourcer=ns.sourcer, use_tmpfile=ns.use_tmpfile)
     # apply results
     env = builtins.__xonsh_env__
     denv = env.detype()
     for k, v in fsenv.items():
         if k in denv and v == denv[k]:
             continue  # no change from original
         env[k] = v
+    # If run in un-safe mode we are sure the command completed correctly,
+    # thus we can remove any env-vars that were unset by the script.
+    if not ns.safe:
+        for k in denv:
+            if k not in fsenv:
+                env.pop(k, None)
```

Comment:
Could we have `source_foreign()` return None if the command somehow failed? Then we could know not to trust the result before removing env-vars. Now the command has to be run with `safe=False` so we can be sure it complated correctly.